### PR TITLE
themes for lavaland

### PR DESCRIPTION
**Pull request by tralezab**
## About The Pull Request

lavaland is sometimes snowland, wasteland or waterworld

## Why It's Good For The Game

lavaland can get a little samey

## Changelog
:cl:
add: snowland
add: wasteland
add: waterworld
/:cl:

### CODE REVIEW

**Comment by actioninja** [comment]
uuh

**Comment by cacogen** [comment]
oh ok

**Comment by tralezab** [comment]
do note: headmins will have to turn this on in the config, switching "lavaland" to "themed"

**Comment by cacogen** [comment]
you may want to post some pictures, maybe expand on how the system works a little

**Comment by Spepbuthewranglesgithub** [comment]
im sure nothing here will conflict with item/structure descriptions at all ever and nothing will ever break.

_this would actually be real nice for spicing things up visually but oh god the descriptions_

**Comment by tralezab** [comment]
yeah i'm going to have to work on some ruin mapping a bit to work them in with the others.
luckily for walls and the like i can mostly use thematic turfs to adjust them on the fly

**Comment by nemvar** [comment]
Does this actually work yet? I only see support for snowland in the code. Shouldn't this have a WIP tag?

**Comment by tralezab** [comment]
it's really close to completion!!

**Comment by trollbreeder** [comment]
Close, but not fully.

**Comment by HomocidlePancakes** [comment]
It changes the mov sprites, riiiiiiight?

**Comment by tralezab** [comment]
Sure! We have snow legion mobs so it could easily happen

**Comment by Farquaar** [comment]
Any screenshots?

**Comment by tralezab** [comment]
should have tuned into my livestream in the discord voice chat general

smh smh well i'll post screenshots of at least one of the worlds i've done (i have a few locally)

**Comment by Qustinnus** [comment]
Yeah this needs screenshots im very skeptical about all of this lol

**Comment by wesoda25** [comment]
Idk how much work it would entail, but if you changed the space background with lavaland in it so that it fit whichever theme, that would be awesome

File: code/game/turfs/open/lava.dm [hunk 1]
```diff
@@ -190,3 +190,100 @@
...
+				if(O.resistance_flags & FREEZE_PROOF)
+					continue
+
+			else if(isliving(buckle_check))
+				var/mob/living/live = buckle_check
+				if("snow" in live.weather_immunities)
+					continue
+
+			L.adjustFireLoss(2)
+			if(L)
```
Comment (Qustinnus) [inline code/game/turfs/open/lava.dm]: what the fuck is this nested ass code arm, use some early returns pls

File: code/__DEFINES/maps.dm [hunk 2]
```diff
@@ -64,11 +75,24 @@ require only minor tweaks.
...
+#define ZTRAITS_LAVAWORLD list(\
+	ZTRAIT_BASETURF = /turf/open/lava/smooth/lava_land_surface, \
+	ZTRAIT_THEMATIC_PLATING = /turf/open/floor/plating/asteroid/basalt/lava_land_surface, \
+	ZTRAIT_THEMATIC_LAVA = /turf/open/lava/smooth/lava_land_surface, \
+	ZTRAIT_WEATHER_ASHSTORM = TRUE )
+	//ZTRAIT_MINERAL_ICON = not included because lavaworld uses the default
+
+#define ZTRAITS_SNOWWORLD list(\
+	ZTRAIT_BASETURF = /turf/open/lava/plasma/lava_land_surface, \
+	ZTRAIT_THEMATIC_PLATING = /turf/open/floor/plating/asteroid/snow/lava_land_surface, \
```
Comment (Qustinnus) [inline code/__DEFINES/maps.dm]: this entire path makes me sad. why is lava_land_surface a child of two types.

File: code/game/turfs/closed/minerals.dm [hunk 3]
```diff
@@ -24,6 +24,9 @@
 /turf/closed/mineral/Initialize()
 	if (!canSmoothWith)
 		canSmoothWith = list(/turf/closed/mineral, /turf/closed/indestructible)
+	var/ztrait = SSmapping.level_trait(z, ZTRAIT_MINERAL_ICON)
```
Comment (Qustinnus) [inline code/game/turfs/closed/minerals.dm]: I can't explain why but the idea of the ztrait changing the icon of the turfs on init doesn't sit right with me.

File: code/game/turfs/open/floor/plating/thematic.dm [hunk 4]
```diff
@@ -0,0 +1,12 @@
+/turf/thematic
+	var/trait
+
+/turf/thematic/New()
```
Comment (Qustinnus) [inline code/game/turfs/open/floor/plating/thematic.dm]: You can't expect to change ~255*255 (excluding the lava/station turfs) turfs to another theme and not expect major lag. Especially if its on New() (Why is it on new in the first place)

**Review by ghost — CHANGES_REQUESTED** [review]
I feel like this is going to mess up lavaland code even more than its already a massive clusterfuck arm, you might want to look into alternative methods, or just refactor the way lavaland works in the first place, because this just makes it more unmaintainable

**Comment by Whoneedspacee** [comment]
not sure how this is going to interact with this thing ive been doing for a while

https://www.youtube.com/watch?v=iB-hkN_wYmY

**Comment by Qustinnus** [comment]
@wesoda25 we just need the sprites

**Comment by trollbreeder** [comment]
Should we let players vote for which mining map they want?

**Comment by optimumtact** [comment]
The real problem here is one of rsc bloat and the impact that has on the bandwidth usage of /tg/

**Comment by Qustinnus** [comment]
I don't really see the purpose either way because all it does is increase maintenance needed for the same content

File: code/__DEFINES/maps.dm (hunk 2 above)
Comment (tralezab) [reply code/__DEFINES/maps.dm]: turf pathing is fucked, this is nothing new

Comment (Qustinnus) [reply code/__DEFINES/maps.dm]: dont make it worse :^(

File: code/game/turfs/open/floor/plating/thematic.dm (hunk 4 above)
Comment (tralezab) [reply code/game/turfs/open/floor/plating/thematic.dm]: mining literally already did this when minerals floyddabbed on New and changed their turf to whatever mineral holds the ore they generated. gobless thats gone

**Comment by tralezab** [comment]
now, we can test the actual cost in game but as a reminder every mineral on the mining zlevel changed turf to another mineral when they loaded
now minerals don't do that and floors and lava do that, so it's either the same load as when lavaland did it or way less

**Comment by Qustinnus** [comment]
The argument is "It'll only freeze the server up as much as it did before" is not really strong in my opinion, that combined with the fact that this runs in New() and not Initialize() which I assume is what mineral turfs did.

**Comment by tralezab** [comment]
icemoon is here tada~

**Comment by wesoda25** [comment]
bro i still want pictures of this

**Comment by tralezab** [comment]
@wesoda25 join voice chat general sometime i'll launch it up

**Pull request closed without merging**